cmd/kube-apiserver/app/testing/testserver.go

```diff
@@ -318,6 +323,14 @@ func StartTestServer(t Logger, instanceOptions *TestServerInstanceOptions, custo
 		return result, err
 	}
 
+	featuregatefs.VisitAll(func(flag *pflag.Flag) {
+		if !flag.Changed {
+			return
+		}
+
+		t.Fatalf("Flag %q mutates global state and should not be used in integration tests. Prefer SetFeatureGateDuringTest from k8s.io/component-base/featuregate/testing.", flag.Name)
```

Comment:
alternatively, could add a snapshot/restore function to featuregate/testing like this:

```go
func SnapshotFeatureGateState(tb testing.TB, gate featuregate.FeatureGate) func() {
	var cleanups []func()

	// Snapshot all gates so their individual values get restored
	for k, v := range gate.(featuregate.MutableFeatureGate).GetAll() {
		if k == "AllAlpha" || k == "AllBeta" {
			continue
		}
		if !v.LockToDefault {
			cleanups = append(cleanups, SetFeatureGateDuringTest(tb, gate, k, gate.Enabled(k)))
		}
	}

	return func() {
		for _, cleanup := range cleanups {
			cleanup()
		}
	}
}
```

invoke that in this function:
```
restoreFeatureGates := featuregatetesting.SnapshotFeatureGateState(...)
```

and call `restoreFeatureGates()` from the `tearDown` function